### DISCUSSION TITLE
How to determine if a 2D molecule is symmetric or asymmetric

### DISCUSSION THREAD

**Discussion by brigreens** [rdkit/rdkit#6229] [Q&A]
Hello, does anyone has any ideas on determining if a 2D molecule (based on the connectivity of atoms, stereochemistry does not matter) is symmetric or asymmetric?

For example, I would want to determine that this molecule is asymmetric: `CCC1=C(SC2=C1SC3=C2N(CC)C4=C3C5=NSN=C5C6=C4N(CC)C7=C6SC8=C7N(CC)C9=C8SC(/C=C%10C(C%11=C(C\%10=C(C#N)/C#N)C=C(F)C(F)=C%11)=O)=C9)/C=C%12C(C%13=C(C\%12=C(C#N)/C#N)C=C(F)C(F)=C%13)=O`

And this molecule is symmetric: `CCC1=C(SC2=C1SC3=C2N(CC)C4=C3C5=NSN=C5C6=C4N(CC)C7=C6SC8=C7SC(/C=C9C(C%10=C(C\9=C(C#N)/C#N)C=C(F)C(F)=C%10)=O)=C8CC)/C=C%11C(C%12=CC(F)=C(C=C%12C\%11=C(C#N)/C#N)F)=O`

Thank you. Any ideas greatly appreciated!

**Comment by greglandrum** [comment]
Hi @brigreens. That depends on what you mean by "symmetric".

One straightforward definition of symmetry is whether or not a molecule has more than one substructure match to itself. In this case both of your molecules are symmetric - this is because in the first moelcule the two =CC(C#N)C#N groups can each map to themselves:
```
In [10]: len(m1.GetSubstructMatches(m1,uniquify=False))
Out[10]: 4

In [11]: len(m2.GetSubstructMatches(m2,uniquify=False))
Out[11]: 8
```
Another definition (which ends up providing the same results a bit quicker) is to use the canonical atom numbering code with tie breaking turned off and to look at the number of unique ranks you get back: if this is less than the number of atoms in the molecule then you have a symmetry). Again, both of your molecules have symmetry in this representation:
```
In [7]: len(set(Chem.CanonicalRankAtoms(m1,breakTies=False)))
Out[7]: 70

In [8]: len(set(Chem.CanonicalRankAtoms(m2,breakTies=False)))
Out[8]: 34

In [9]: m1.GetNumAtoms()
Out[9]: 74
```

**Reply by brigreens** [reply]
Hi @greglandrum, thank you for the reply. 
The definition of "symmetric" I am aiming for is having a vertical mirror plane or center of inversion. I would want a way for m1 to return as asymmetric and m2 to return as symmetric. In both of your examples, as, you mentioned, they would both return as symmetric. There are probably ways to do this with other programs using point groups, however, I am most comfortable with RDKit so I would love to know if there is a way to do this with it.

**Reply by queliyong** [reply]
Hi,
I find that when a mol is symmetric, len(set(Chem.CanonicalRankAtoms(m2,breakTies=False))) is far less than len(Chem.CanonicalRankAtoms(m2,breakTies=False)), and mostly less than half of latter one. The conclusion works for your two mols,and several other examples. It would be appreciated if any better answers provided.

**Comment by dehaenw** [comment]
interesting problem. in line with the canonical rank approach, i'd suggest the following approach:
- create a morganlike identifier for each unique atom environment. the environment is the whole molecule in this case in order to prevent locally but not globally identical atom neighborhoods
- create identifier for bonds based on its begin and end atom identifiers
- delete all the unique bonds, its associated atoms and atoms that have an identifier that occurs only once in the molecule (if there are no uniques of these probably a check for other higher symmetries like 3- or 4-fold should be done, if there are only uniques the molecule doesnt have a chance of being symmetric)
-if the resulting fragments contain only identical pairs, the molecule is symmetrical


here is a quick prototype of this idea. a little bit messy, but it works on all the test cases i tried. 
```python
from rdkit import Chem
from collections import Counter

def is_mol_symmetric(mol):
    symmetric=False
    
    #generate morgan like invariants
    atoms=mol.GetAtoms()
    bonds=mol.GetBonds()
    neighbors={i:[n.GetIdx() for n in a.GetNeighbors()] for i,a in enumerate(atoms)}
    atom_invariants=[a.GetAtomicNum() for a in atoms]
    for i in range(len(atoms)-1):
        new_atom_invariants=[]
        for j in range(len(atoms)):
            new_atom_invariants.append(hash(tuple(sorted([atom_invariants[idx] for idx in neighbors[j]]))))
        atom_invariants=new_atom_invariants
    bond_invariants=[atom_invariants[b.GetBeginAtomIdx()]|atom_invariants[b.GetEndAtomIdx()] for b in bonds]

    #print unique ids:
    unique_b_idx=[]
    unique_a_idx=[]
    acounter=Counter(atom_invariants)
    bcounter=Counter(bond_invariants)
    for inv in acounter:
        if acounter[inv]==1:
            unique_a_idx.append(atom_invariants.index(inv))
    for inv in bcounter:
        if bcounter[inv]==1:
            unique_b_idx.append(bond_invariants.index(inv))

    #check the count of uniques
    if len(unique_a_idx+unique_b_idx)==0:
        print("probably higher symmetry?")
    elif len(unique_a_idx)==len(atoms):
        pass #only unique atoms: totally not symmetric"
    else:
        
        #delete the unique bonds and atoms
        newmol=Chem.RWMol(mol)
        deleteextraatoms=[]
        for idx in unique_b_idx:
            a1=bonds[idx].GetBeginAtomIdx()
            a2=bonds[idx].GetEndAtomIdx()
            newmol.RemoveBond(a1,a2)
            deleteextraatoms+=[a1,a2]
        
        deleteatoms=sorted(list(set(unique_a_idx+deleteextraatoms)))[::-1]
        for idx in deleteatoms:
            newmol.RemoveAtom(idx)
        
        #remove aromatic flags and sanitize
        for a in newmol.GetAtoms():
            if (not a.IsInRing()) and a.GetIsAromatic():
                a.SetIsAromatic(False)
        Chem.SanitizeMol(newmol)


        #check if there are only pairs of fragments and if they are identical
        frags=Chem.GetMolFrags(newmol,asMols=True)
        smis=[Chem.MolToSmiles(f) for f in frags]
        smicounter=Counter(smis)
        symmetric=True
        for k in smicounter:
            if smicounter[k]!=2:
                symmetric=False
    return symmetric
```

a few test molecules that should give `True`: `["CCCOCOCCC","FC1CC2C1CC2F","c1ccccc1C#Cc1ccccc1","CCCN(C)CCC","o1ccc2c1cco2"]`

gratifyingly, when tested on the molecules from the original question, it correctly picks out the mirror symmetric one.

```python
smis=["CCC1=C(SC2=C1SC3=C2N(CC)C4=C3C5=NSN=C5C6=C4N(CC)C7=C6SC8=C7N(CC)C9=C8SC(/C=C%10C(C%11=C(C\%10=C(C#N)/C#N)C=C(F)C(F)=C%11)=O)=C9)/C=C%12C(C%13=C(C\%12=C(C#N)/C#N)C=C(F)C(F)=C%13)=O", "CCC1=C(SC2=C1SC3=C2N(CC)C4=C3C5=NSN=C5C6=C4N(CC)C7=C6SC8=C7SC(/C=C9C(C%10=C(C\9=C(C#N)/C#N)C=C(F)C(F)=C%10)=O)=C8CC)/C=C%11C(C%12=CC(F)=C(C=C%12C\%11=C(C#N)/C#N)F)=O"]
for smi in smis:
    mol=Chem.MolFromSmiles(smi)
    display(mol)
    print(is_mol_symmetric(mol))
```
returns
![screenshotsymmetry](https://github.com/rdkit/rdkit/assets/66372095/beef841b-a069-456d-aca2-638234cc2c7d)
